CLI tests audit and update + multiple bug fixes in data context profile_datasource and CLI logic that the tests helped uncover

eugmandel opened the pull request:
Audited, updated, fixed, added CLI tests.

This helped uncover several bugs, weird/underimplemented behaviors in the data context and in the cli/datasources.

The only CLI tests that have "xfails" now are test_init_pandas and test_init_sqlite.

This PR is ready to be reviewed. All tests pass on Py3.

eugmandel committed:
Multiple fixes in data context profile_datasource, updated the pandas datasource cli test - passing, sqlite datasource cli tests do not pass yet

pep8speaks commented:
Thank you for submitting to Great Expectations!

* In the file [`great_expectations/cli/datasource.py`](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py):

> [Line 132:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L132): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 120 characters)
> [Line 152:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L152): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 120 characters)
> [Line 773:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L773): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (123 > 120 characters)
> [Line 889:1](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L889): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (4)
> [Line 892:40](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L892): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 893:104](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L893): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 903:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L903): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 120 characters)
> [Line 1111:79](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L1111): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1114:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L1114): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 120 characters)
> [Line 1115:86](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L1115): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1187:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/cli/datasource.py#L1187): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (152 > 120 characters)



* In the file [`great_expectations/data_context/data_context.py`](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/data_context/data_context.py):

> [Line 1000:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/data_context/data_context.py#L1000): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (130 > 120 characters)
> [Line 1032:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/data_context/data_context.py#L1032): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (129 > 120 characters)
> [Line 1034:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/great_expectations/data_context/data_context.py#L1034): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 120 characters)



* In the file [`tests/cli/test_datasource_sqlite.py`](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/cli/test_datasource_sqlite.py):

> [Line 327:98](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/cli/test_datasource_sqlite.py#L327): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 328:36](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/cli/test_datasource_sqlite.py#L328): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 328:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/cli/test_datasource_sqlite.py#L328): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (134 > 120 characters)

* In the file [`tests/cli/test_init.py`](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/cli/test_init.py):

> [Line 160:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/cli/test_init.py#L160): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 120 characters)





* In the file [`tests/cli/test_suite.py`](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/cli/test_suite.py):

> [Line 25:1](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/cli/test_suite.py#L25): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 142:1](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/cli/test_suite.py#L142): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1



* In the file [`tests/profile/test_profile.py`](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/profile/test_profile.py):

> [Line 340:40](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/profile/test_profile.py#L340): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to True should be 'if cond is True:' or 'if cond:'
> [Line 342:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/profile/test_profile.py#L342): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 120 characters)
> [Line 344:1](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/profile/test_profile.py#L344): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 354:121](https://github.com/great-expectations/great_expectations/blob/a902c6fc5dd1f059c91565e4b483c58ddcc26e8b/tests/profile/test_profile.py#L354): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (140 > 120 characters)

This linter comment has been updated based on the most recent commit.

##### Comment last updated at 2020-02-05 19:54:20 UTC

eugmandel committed:
Merge remote-tracking branch 'origin/0.9.0' into eugene/0.9.0_100

eugmandel committed:
Adjusted/fixed tests in tests/cli/test_datasource_sqlite.py - all are passing now.
Had to improve how data context's profile_datasource and cli/datasource's profile_datasource handle the generators.

eugmandel committed:
Merge remote-tracking branch 'origin/0.9.0' into eugene/0.9.0_100

eugmandel committed:
merge

eugmandel committed:
cli/test_init adjusted + an issue that the test uncovered

eugmandel committed:
merge

eugmandel committed:
changes in pytest markings

eugmandel committed:
The view arg should be passed to build_docs

eugmandel committed:
a tiny change in the expected string

eugmandel committed:
Fix in the expected string

eugmandel committed:
Fixed the logic and the test for the case of profile_datasource getting datasource_name that does not correspond to any datasource in the context

eugmandel committed:
Improved the logic of cli datasource get_batch to support both generators and file paths + updated the tests

eugmandel committed:
Fix: generator_name was passed to the data context method

Aylr commented on great_expectations/cli/init.py:
```diff
@@ -101,7 +100,7 @@ def init(target_directory, view):
         if context.list_expectation_suite_keys():
             if click.confirm(BUILD_DOCS_PROMPT, default=True):
-                build_docs(context)
+                build_docs(context, view=view)
```
Great catch.

Aylr commented on tests/cli/test_datasource_sqlite.py:
```diff
@@ -130,6 +134,39 @@ def test_cli_datasource_profile_answering_no(
+
+    stdout = result.output
+    print(stdout)
+    assert result.exit_code == -1
```
typically posix exit codes are positive - this should probably be `sys.exit(1)` for general error.

eugmandel replied:
Agreed - it is gone.

Aylr commented on tests/cli/test_datasource_sqlite.py:
```diff
@@ -260,38 +297,42 @@ def test_cli_datasource_profile_with_data_asset_and_additional_batch_kwargs_shou
-    assert isinstance(result.exception, DataContextError)
+    validation = validations_store.get(validation_keys[0])
+    assert validation.meta["expectation_suite_name"] == "wow_a_datasource.default.main.titanic.BasicDatasetProfiler"
+    assert validation.success is False
```
Maybe I don't understand `validation.success` in a profiler - should this really be False?

eugmandel replied:
Some expectation types are not implemented in SQLAlchemy datasource - that's why validation.success is False.

Aylr commented on tests/cli/test_init.py:
```diff
@@ -118,7 +118,7 @@ def test_cli_init_on_existing_project_with_no_uncommitted_dirs_answering_no_to_f
 
 
-@pytest.mark.xfail(reason="failing")
+# @pytest.mark.xfail(reason="failing")
```
let's remove this

eugmandel replied:
Addressed.

Aylr commented on tests/cli/test_init.py:
```diff
@@ -158,11 +158,10 @@ def test_cli_init_on_complete_existing_project_all_uncommitted_dirs_exist(
 
-@pytest.mark.xfail
-def test_cli_init_connection_string_non_working_mssql_connection_instructs_user_and_leaves_entries_in_config_files_for_debugging(
+# @pytest.mark.xfail
```
let's remove this

eugmandel replied:
Addressed.

Aylr commented on tests/cli/test_init.py:
```diff
@@ -182,7 +181,7 @@ def test_cli_init_connection_string_non_working_mssql_connection_instructs_user_
     assert "Attempting to connect to your database. This may take a moment" in stdout
     assert "Cannot connect to the database" in stdout
-    assert "Database Error: No module named 'pymssql'" in stdout
+    # assert "Database Error: No module named 'pymssql'" in stdout
```
let's remove this

eugmandel replied:
Addressed.

Aylr commented on tests/cli/test_init_pandas.py:
```diff
@@ -229,8 +228,8 @@ def initialized_project(tmp_path_factory):
-@pytest.mark.skip(reason="failing")
-def test_init_on_existing_project_with_multiple_datasources_exist_do_nothing(
+# @pytest.mark.skip(reason="failing")
+def x(
```
This is probably not intentional

eugmandel replied:
A weird cut and paste slip - thank you! Addressed.

Aylr commented on tests/cli/test_init_pandas.py:
```diff
@@ -289,7 +287,7 @@ def test_init_on_existing_project_with_datasource_with_existing_suite_offer_to_b
 
 
-@pytest.mark.skip(reason="failing")
+# @pytest.mark.skip(reason="failing")
```
let's remove this

eugmandel replied:
Addressed.

Aylr commented on tests/cli/test_suite.py:
```diff
@@ -22,26 +22,97 @@ def test_suite_help_output(caplog,):
     assert result.exit_code == 0
     assert "Select data source" in stdout
-    assert "Which data would you like to use" in stdout
+    assert "Which data asset would you like to use" in stdout
```
Nitpick - this introduces data asset language into the CLI. While it is not technically in the init flow, it is a new term.

eugmandel replied:
Agreed - we said we would not introduce this term out of nowhere - addressed.

Aylr commented on tests/cli/test_suite.py:
```diff
@@ -68,25 +139,29 @@ def test_suite_new_without_suite_name_argument(
     assert result.exit_code == 0
     assert "Select data source" in stdout
-    assert "Which data would you like to use" in stdout
+    assert "Which data asset would you like to use" in stdout
```
same as above

eugmandel replied:
Addressed.

Aylr commented on great_expectations/cli/datasource.py:
```diff
@@ -957,14 +974,6 @@ def _load_file_from_filesystem_as_data_asset(context, datasource_name,
 
     path = os.path.abspath(path)
 
-    # if generator_asset is None:
-    #     generator_asset = os.path.splitext(os.path.basename(path))[0]
-    #     generator_asset = click.prompt(
-    #         msg_prompt_data_asset_name,
-    #         default=generator_asset,
-    #         show_default=True
-    #     )
-
```
❤️

Aylr commented on great_expectations/cli/init.py:
```diff
@@ -72,7 +72,6 @@ def init(target_directory, view):
             try:
                 context = DataContext(ge_dir)
                 cli_message(PROJECT_IS_COMPLETE)
-                exit(0)
```
While this is technically redundant (click exits w/ status 0 if a function returns nothing), it does make this easier to read. Could we leave this in?

eugmandel replied:
We don't want to exit here, because the next step (outside of all the if/else) is to offer to build docs.

jcampbell commented on tests/cli/test_init_pandas.py:
```diff
@@ -229,8 +228,8 @@ def initialized_project(tmp_path_factory):
-@pytest.mark.skip(reason="failing")
-def test_init_on_existing_project_with_multiple_datasources_exist_do_nothing(
+# @pytest.mark.skip(reason="failing")
+def x(
```
??

eugmandel replied:
A typo - thank you! Addressed.

Aylr commented on great_expectations/data_context/data_context.py:
```diff
@@ -927,76 +929,108 @@ def profile_datasource(self,
             When success = False, the error details are under "error" key
         """
 
+        datasource = self.get_datasource(datasource_name)
```
This variable isn't used anywhere

eugmandel replied:
We don't need the datasource object, but this line serves to check if the datasource by the name passed as an arg exists and raise an error if it does not. Added a comment to the code.

Aylr commented on great_expectations/data_context/data_context.py:
```diff
@@ -927,76 +929,108 @@ def profile_datasource(self,
+
+        data_asset_names_dict = self.get_available_data_asset_names(datasource_name)
 
         # KeyError will happen if there are no generators
```
this comment may not longer be relevant

eugmandel replied:
Removed the comment.

Aylr commented on great_expectations/data_context/data_context.py:
```diff
@@ -927,76 +929,108 @@ def profile_datasource(self,
+                "No Data Assets found in Datasource {}. Used generator: {}.".format(
                     datasource_name,
-                    list(datasource.keys()))
+                    generator_name)
```
This all makes sense after considerable study. However I don't see testing for all the end states.

eugmandel replied:
You are right! Adding tests in a separate PR.

Aylr commented on great_expectations/cli/datasource.py:
```diff
@@ -125,6 +125,12 @@ def datasource_list(directory):
 
 @datasource.command(name="profile")
 @click.argument('datasource_name', default=None, required=False)
+@click.option(
+    "--generator_name",
+    "-g",
+    default=None,
+    help="The name of the batch kwarg generator configured in the datasource. The generator will list data assets in the datasource"
+)
```
While this makes sense, it is a new feature that definitely introduces a new dimension of combinations that appear yet unproven with tests.

eugmandel replied:
You are right! Adding tests in a separate PR.

eugmandel committed:
Fixed a typo that renamed a testing method to x

eugmandel committed:
Addressing PR comments

eugmandel committed:
Addressing PR comments
